File: src/110-balanced-binary-tree/solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
  return true unless root

  dfs = lambda do |node|
    return 0 unless node

    l, r = dfs.call(node.left), dfs.call(node.right)
    return - 1 if l == -1 || r == -1 || (l - r).abs > 1

    1 + [l, r].max
  end
  
  dfs.call(root) > 0
end

File: src/110-balanced-binary-tree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {
    if(!root) return true;

    const dfs = (node: TreeNode): number => {
        if(!node) return 0;

        let l: number = dfs(node.left), r: number = dfs(node.right);

        if(l === -1 || r === -1 || Math.abs(l - r) > 1) return -1;

        return 1 + Math.max(l, r);
    } 

    return dfs(root) > 0;
};
